### PR TITLE
Update on-boarding UI

### PR DESCRIPTION
**Pull request by Skrilltrax**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## :scroll: Description
Onboarding UI changes.

## :bulb: Motivation and Context
Our previous onboarding UI was a bunch of dialogs inside a single activity. This new UI makes it easy to extend the onboarding activity to add new features.

## :green_heart: How did you test it?
Tested all the onboarding options.

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code
- [x] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

## :camera_flash: Screenshots / GIFs

### CODE REVIEW

**Comment by Skrilltrax** [comment]
I'm following [Buffer's package structure](https://buffer.com/resources/android-rethinking-package-structure/) here. We should fix all our packages like this once we are ready to push V2.

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -153,6 +153,7 @@
             android:configChanges="orientation|keyboardHidden"
             android:theme="@style/DialogLikeTheme"
             android:windowSoftInputMode="adjustNothing" />
+        <activity android:name="com.zeapo.pwdstore.ui.onboarding.activity.OnboardingActivity" />
```
Comment (msfjarvis) [inline app/src/main/AndroidManifest.xml]:
```suggestion
        <activity android:name=".ui.onboarding.activity.OnboardingActivity" />
```

Comment (msfjarvis) [reply app/src/main/AndroidManifest.xml]: Also, why the name re-use?

File: app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/CloneFragment.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        super.onCreateView(inflater, container, savedInstanceState)
+        binding = FragmentCloneBinding.inflate(inflater)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.btnCloneRemote.setOnClickListener {
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/CloneFragment.kt]: Drop the `btn` prefix, it doesn't hold much value in a strongly-typed ViewBinding world.

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -33,7 +33,7 @@
             android:label="@string/app_name" />
 
         <activity
-            android:name=".OnboardingActivity"
+            android:name="com.zeapo.pwdstore.ui.onboarding.activity.OnboardingActivity"
```
Comment (msfjarvis) [inline app/src/main/AndroidManifest.xml]:
```suggestion
            android:name=".ui.onboarding.activity.OnboardingActivity"
```

File: app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt [hunk 4]
```diff
@@ -1,56 +1,53 @@
 /*
- * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
- * SPDX-License-Identifier: GPL-3.0-only
+ * Copyright © 2019-2020 The Android Password Store Authors. All Rights Reserved.
+ *  SPDX-License-Identifier: GPL-3.0-only
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt]: leave this as-is.

File: app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt [hunk 5]
```diff
@@ -208,4 +184,16 @@ class OnboardingActivity : AppCompatActivity() {
         }
         createRepository()
     }
+
+    private fun createPermGrantedAction(block: () -> Unit): ActivityResultLauncher<String> =
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt]: Is the type necessary?

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt [hunk 6]
```diff
@@ -40,21 +43,14 @@ import org.eclipse.jgit.revwalk.RevCommit
...
-fun String.splitLines(): Array<String> {
-    return split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
-}
+fun String.splitLines(): Array<String> =
+    split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
 
-fun String.base64(): String {
-    return Base64.encodeToString(encodeToByteArray(), Base64.NO_WRAP)
-}
+fun String.base64(): String = Base64.encodeToString(encodeToByteArray(), Base64.NO_WRAP)
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt]: Revert these

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt [hunk 7]
```diff
@@ -188,3 +188,39 @@ val RevCommit.time: Date
...
+fun FragmentManager.performSharedElementTransaction(destinationFragment: Fragment, views: List<View>, @IdRes containerViewId: Int = android.R.id.content) {
+    this.commit {
+        beginTransaction()
+        for (view in views) {
+            addSharedElement(view, view.transitionName)
+        }
+        addToBackStack(destinationFragment.tag)
+        replace(containerViewId, destinationFragment)
+    }
+}
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt]: Move them to a separate `FragmentExtensions` file, things are getting pretty wild in here. Need to redistribute some more but that can go in a separate PR.

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt [hunk 8]
```diff
@@ -165,9 +166,8 @@ val Context.autofillManager: AutofillManager?
 val Context.keyguardManager: KeyguardManager
     get() = getSystemService()!!
 
-fun File.isInsideRepository(): Boolean {
-    return canonicalPath.contains(getRepositoryDirectory().canonicalPath)
-}
+fun File.isInsideRepository(): Boolean =
+    canonicalPath.contains(getRepositoryDirectory().canonicalPath)
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt]: Revert this

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt [hunk 9]
```diff
@@ -135,9 +131,14 @@ suspend fun FragmentActivity.commitChange(
...
-    return ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED
-}
+fun FragmentActivity.isPermissionGranted(permission: String): Boolean =
+    ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED
+
+
+fun Fragment.isPermissionGranted(permission: String): Boolean =
+    ContextCompat.checkSelfPermission(requireActivity(), permission) == PackageManager.PERMISSION_GRANTED
+
+fun Fragment.finish() = requireActivity().finish()
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt]: Move to `FragmentExtensions` along with the extensions at the bottom of this file.

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt [hunk 10]
```diff
@@ -135,9 +131,14 @@ suspend fun FragmentActivity.commitChange(
     }.execute()
 }
 
-fun FragmentActivity.isPermissionGranted(permission: String): Boolean {
-    return ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED
-}
+fun FragmentActivity.isPermissionGranted(permission: String): Boolean =
+    ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt]: Revert this

File: app/src/main/res/layout/fragment_repo_location.xml [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+    <TextView
+        android:id="@+id/tv_repo_location"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="48dp"
+        android:text="@string/repository_n_location"
+        android:textStyle="bold"
+        android:textColor="@color/color_control_normal"
+        android:textAppearance="@style/TextAppearance.AppCompat.Display1"
```
Comment (msfjarvis) [inline app/src/main/res/layout/fragment_repo_location.xml]: Let's try to stick with `MaterialComponents` text styling

**Review by msfjarvis — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt (hunk 5 above)
Comment (Skrilltrax) [reply app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt]: `ActivityResultContracts.RequestPermission()` always return `ActivityResultLauncher<String>`

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt]: yeah but do you need to explicitly set the type? I'd assume type inference catches it

Comment (Skrilltrax) [reply app/src/main/java/com/zeapo/pwdstore/ui/onboarding/fragments/RepoLocationFragment.kt]: Yeah, it does, AS was showing error for me earlier IDK why. Fixed now.

**Comment by msfjarvis** [comment]
Removed reformat commit since it added too many changes to an already big PR, it can go in separately later.

**Review by msfjarvis — APPROVED** [review]
Everything appears to work in my testing

**Comment by fmeum** [comment]
While updating the Wiki entries, I noticed that it doesn't seem to be possible anymore to replace an existing SSH key from the on-boarding UI. Should we reintroduce the access to settings or just reset it if the repository is deleted?

**Comment by msfjarvis** [comment]
> While updating the Wiki entries, I noticed that it doesn't seem to be possible anymore to replace an existing SSH key from the on-boarding UI. Should we reintroduce the access to settings or just reset it if the repository is deleted?

I'm not a fan of either option but adding a settings option feels like the less-likely-to-be-annoying choice w.r.t. UX.

**Comment by fmeum** [comment]
> > While updating the Wiki entries, I noticed that it doesn't seem to be possible anymore to replace an existing SSH key from the on-boarding UI. Should we reintroduce the access to settings or just reset it if the repository is deleted?
> 
> I'm not a fan of either option but adding a settings option feels like the less-likely-to-be-annoying choice w.r.t. UX.

Agreed. @Skrilltrax Is this something you could add before the release? I will update the remaining Wiki entries.

**Comment by msfjarvis** [comment]
> > > While updating the Wiki entries, I noticed that it doesn't seem to be possible anymore to replace an existing SSH key from the on-boarding UI. Should we reintroduce the access to settings or just reset it if the repository is deleted?
> > 
> > 
> > I'm not a fan of either option but adding a settings option feels like the less-likely-to-be-annoying choice w.r.t. UX.
> 
> Agreed. @Skrilltrax Is this something you could add before the release? I will update the remaining Wiki entries.

He's got a fair bit of stuff going on at work, I'll do it.

**Comment by fmeum** [comment]
> > > > While updating the Wiki entries, I noticed that it doesn't seem to be possible anymore to replace an existing SSH key from the on-boarding UI. Should we reintroduce the access to settings or just reset it if the repository is deleted?
> > > 
> > > 
> > > I'm not a fan of either option but adding a settings option feels like the less-likely-to-be-annoying choice w.r.t. UX.
> > 
> > 
> > Agreed. @Skrilltrax Is this something you could add before the release? I will update the remaining Wiki entries.
> 
> He's got a fair bit of stuff going on at work, I'll do it.

Thanks! The Wiki entries should now reflect the flow for 1.12.0. I have also made them recommend in-app generation of SSH keys.

**Comment by msfjarvis** [comment]
> > > > > While updating the Wiki entries, I noticed that it doesn't seem to be possible anymore to replace an existing SSH key from the on-boarding UI. Should we reintroduce the access to settings or just reset it if the repository is deleted?
> > > > 
> > > > 
> > > > I'm not a fan of either option but adding a settings option feels like the less-likely-to-be-annoying choice w.r.t. UX.
> > > 
> > > 
> > > Agreed. @Skrilltrax Is this something you could add before the release? I will update the remaining Wiki entries.
> > 
> > 
> > He's got a fair bit of stuff going on at work, I'll do it.
> 
> Thanks! The Wiki entries should now reflect the flow for 1.12.0. I have also made them recommend in-app generation of SSH keys.

The Wiki changes look good, I've created #1119 to add the settings button.

**Pull request merged**